File: standard_problems/Arrays&Hashing/move_all_zeros_at_the_end.py
# Give an integer array, move all zeros present in it 
# to the end. The solution should maintain the relative order of the items
# in the array.

# Keet track of the zeros, and swap it with the next non zero element in the array
# Time: O(N)
# Space: O(1)

def MoveZeros(arr):

    zero_pos = 0

    for i in range(len(arr)):
        if arr[i]!=0:
            arr[zero_pos], arr[i] = arr[i], arr[zero_pos]
            zero_pos += 1
    return arr

if __name__ == "__main__":
    MoveZeros([6, 0, 8, 2, 3, 0, 4, 0, 1])

File: standart_problems/Arrays_and_Hashing/move_all_zeros_at_the_end.cpp
#include <iostream>
using namespace std;

void swap(int arr[], int i, int j)
{
    int tmp = arr[i];
    arr[i] = arr[j];
    arr[j] = tmp;

}
void move_zeros(int arr[], int n)
{

    int k = n-1;
    
    for (int i=n-1; i >=0; i--)
    {
        if (arr[i] == 0){
            swap(arr, k, i);
            k --;

        }
    }
}

int main()
{
    int arr[] = {6, 0, 8, 2, 3, 0, 4, 0, 1};
    int n = sizeof(arr)/sizeof(arr[0]);

    move_zeros(arr, n);

    for (int i=0; i<n; i++){
        cout<<arr[i]<< "  ";

    }
    cout<<endl;
}